### PR TITLE
Amplitude event provenance as a committed CSV (DATA-2014, folds DATA-2007)

### PR DESCRIPTION
**Pull request by datawithtristan**
# PR: Amplitude event provenance as a committed CSV (DATA-2014, folds DATA-2007)

Branch: `data-2014-provenance-refresh` → `main`. Closes DATA-2014. Supersedes and **closes PR #514 without merging** (its DATA-2007 commits are included here).

## What this delivers

A self-contained tool that builds and maintains the Amplitude event git-provenance dataset, and the dataset itself committed to the repo:

- `analytics/data/amplitude_event_provenance.csv` — one row per Amplitude `event_type` (434 rows), recording when it was instrumented / retired in omni, the attributing PR, and the last code change. Derived purely from omni git history.
- `analytics/data/amplitude_event_provenance_state.json` — the SHA watermark for incremental refresh.
- `analytics/lib/amplitude_event_provenance_backfill.py` — the pipeline (+ full test suite).

## Why a CSV (revised approach)

DATA-2014 originally planned a Databricks table refreshed by an Astro DAG under a service principal. That is superseded: the dataset is now a CSV committed in this repo, so its git history is the audit log of when each event's provenance changed, and there is no service-principal write grant to provision. A committed CSV does not fit a plain Astro DAG (which cannot easily commit to git); the scheduling mechanism is being designed separately (likely the Claude Code Desktop scheduler or a scheduled CI job that opens a PR with the refreshed CSV).

## How it works

- Walks `git log -p origin/develop` over `packages/gp-webapp` + `packages/gp-api`, anchoring on the authoritative event universe (`airbyte_source.amplitude_taxonomy_event_type`, ~434 events synced from Amplitude Govern — the one Databricks read the job makes). Earliest net-add = instrumented, latest net-remove = retired.
- PR attribution is pure git via the merge-commit ancestry walk (no GitHub API — it returns nothing for this repo).
- Single auto-detecting entry point: no state file → full backfill; state file present → walk `last_sha..origin/develop`, update only the events that changed (fresh `updated_at`), carry the rest forward verbatim, rewrite the full sorted CSV, advance the watermark.
- Output schema: `event_type, event_type_slug, instrumented_{commit,pr,date}, retired_{commit,pr,date}, last_code_change_date, updated_at`. `code_status` / `still_in_code` are dropped — derivable from the date null-pattern at read time.

## Folds DATA-2007

This branch already contains the DATA-2007 backfill commits and is based on current `main`, so the single PR delivers backfill + incremental refresh + the CSV pivot together. PR #514 should be closed without merging.

## Known limitation

A refresh only adds/updates events that changed in the window; a brand-new universe event whose instrumentation predates the watermark, or an event removed from the universe, is not reflected until a full backfill (delete the state file). The job logs a warning when universe events are missing from the CSV. This is the accepted design (the full backfill is the resync path).

## Testing

72 unit tests pass (`cd analytics && uv run pytest tests/test_amplitude_event_provenance_backfill.py`); ruff + ruff-format clean. The committed CSV was generated by a real backfill against the omni checkout (434 rows: present=370, removed=21, not_found_in_code=43).

## Out of scope / follow-ups

- Scheduling mechanism (separate design).
- Rebuilding `int__amplitude_event_taxonomy` to source from `airbyte_source.amplitude_taxonomy_event_type` (separate dbt PR, DATA-2005-adjacent).
- DATA-2005 reconciles against this CSV at check time; DATA-2006 Govern write-back unchanged.

---

> [!NOTE]
> **Low Risk**
> Analytics-only artifact and offline git/Databricks reads; no runtime app or auth changes. Incremental refresh can drift from the taxonomy until a full backfill (documented with warnings).
> 
> **Overview**
> Adds a **git-provenance dataset** for Amplitude events: a committed `analytics/data/amplitude_event_provenance.csv` (~434 rows) plus a JSON SHA watermark, replacing the earlier Databricks-table approach.
> 
> **New pipeline** `amplitude_event_provenance_backfill.py` loads the event universe from Databricks (`airbyte_source.amplitude_taxonomy_event_type`), walks `git log -p` on omni `origin/develop` over `packages/gp-webapp` and `packages/gp-api`, and attributes earliest net-add / latest net-remove per literal. PR numbers are filled via merge-commit ancestry (no GitHub API). **Incremental refresh** uses the watermark (`last_sha..ref`), updates only changed events, and rewrites the sorted CSV; missing state triggers a full backfill.
> 
> **`databricks_conn`** gains `get_connection()` so the job can hold a cursor; `run_query` delegates to it.
> 
> **Tests** cover parsing, merge logic, CSV round-trip, refresh/backfill orchestration, and git edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 65dd3504bf0e7f5fdf049c5b46a44a0e62613ff1. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: analytics/lib/amplitude_event_provenance_backfill.py [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+      - last_change: always the window's -- the event net-changed in the window.
+    """
+    if existing_row and existing_row.get("instrumented_commit"):
+        instrumented = _pseudo_commit(
+            existing_row["instrumented_commit"],
+            existing_row["instrumented_pr"],
+            existing_row["instrumented_date"],
+        )
+    else:
+        instrumented = new_entry["instrumented"]
```
[the remark this answers is not in the document]
Comment (datawithtristan) [reply analytics/lib/amplitude_event_provenance_backfill.py:321]: Good catch, fixed in e6df269. The blank `instrumented_*` was ambiguous: a genuinely new event vs. an event whose true instrumentation predates the `--since` window (a `not_found_in_code` row that is actually present in code). The refresh now greps the affected events at the watermark commit and only adopts the window's net-add when the event was absent before the window; for an event already present at the watermark it leaves instrumentation null rather than stamping a false, too-recent date. Covered by two new tests (`test_merge_predates_window_event_does_not_take_window_instrumented` and `test_run_refresh_does_not_fabricate_instrumented_for_predates_window_event`).

**Comment by datawithtristan** [comment]
delegate review

**Comment by datawithtristan** [comment]
delegate review

**Comment by datawithtristan** [comment]
delegate review

File: analytics/lib/amplitude_event_provenance_backfill.py [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+
+    ``-F`` treats each event as a literal; ``-h`` drops filenames. git grep exits 1
+    when nothing matches (an empty dump), which is not an error here.
+    """
+    patterns: list[str] = []
+    for e in events:
+        patterns += ["-e", e]
+    argv = ["git", "-C", root, "grep", "-F", "-h", "--no-color", *patterns, ref, "--", *paths]
+    proc = subprocess.run(argv, capture_output=True, text=True)
+    return proc.stdout
```
[the remark this answers is not in the document]
Comment (datawithtristan) [reply analytics/lib/amplitude_event_provenance_backfill.py:478]: Fixed in 058200c. `git_grep_present_text` now allows exit 1 (no matches) but raises `CalledProcessError` on 2+, matching the guard in `run_git_log`. A fatal git error can no longer masquerade as an empty dump and silently write every event as removed. Added three tests covering the fatal-error, no-match, and match paths (77 passing).

delegate review

**Comment by datawithtristan** [comment]
delegate review

**Comment by datawithtristan** [comment]
delegate review
